src/helpers.lisp

```diff
@@ -381,3 +381,26 @@ expected to have the following format:
        (,c-fn (c-struct retval)
               ,@(expand-c-fun-args args))
        retval)))
+
+(defmacro define-print-object (type methods)
+  "Defines PRINT-OBJECT for TYPE. METHODS should contain function names which
+will be called with the object as the parameter. The name of the method and
+it's result will be printed in the output as `:METHOD (method obj)`. The first
+call to PRINT-OBJECT will use PRINT-UNREADABLE-OBJECT, while next-methods will
+only print to the same object.
+
+E.g. (define-print-object sometype (x y)) will print something like `#<SOMETYPE :x 10 :y 20>`"
+  `(defmethod print-object ((obj ,type) out)
+     (let* ((class (find-class ',type))
+            (top (car (closer-mop:compute-class-precedence-list (find-class (type-of obj)))))
+            (print-methods (lambda () (dolist (method ',methods)
+                                   (let ((value (handler-case (funcall (symbol-function method) obj)
+                                                  (unbound-slot () "#<unbound-slot>"))))
+                                     (format out ":~A ~A " method value))))))
+       (if (eq top class)
+           (print-unreadable-object (obj out :type t :identity t)
+             (funcall print-methods)
+             (call-next-method))
+           (progn
+             (funcall print-methods)
+             (call-next-method))))))
```

Comment:
If you filter this through `compute-class-precedence-list` or similar and prevent calling the next method for `standard-object`, you'll avoid printing the normal object at the end of the string (unless you wanted that for some reason). That will reduce some of the verbosity.

Reply:
I don't really understand the entire class hierarchy in CL, so I didn't dare do too much here. I notice e.g. `font-asset` has the following list: `(#<STANDARD-CLASS CLAYLIB:FONT-ASSET> #<STANDARD-CLASS CLAYLIB::GAME-ASSET>
 #<STANDARD-CLASS COMMON-LISP:STANDARD-OBJECT>
 #<SB-PCL::SLOT-CLASS SB-PCL::SLOT-OBJECT> #<SB-PCL:SYSTEM-CLASS COMMON-LISP:T>)`

Should we filter away only `standard-object`, and print the last two, or should we stop execution once we get to `standard-object`..?

Reply:
Yeah, the idea is to skip over `standard-object` and everything after it. That was the purpose of calling `(cdddr (reverse ...))` on the class precedence list. Unfortunately I don't think there's a simpler way of saying "if the next method would be called on `standard-object` then don't call it" without using that class precedence list.

Reply:
I see. With that patch, it prints: `#<FONT-ASSET :CHARS NIL :SIZE 0 :GLYPH-COUNT 0 :GLYPH-PADDING 0 :TEXTURE #<RL-TEXTURE :ID 0 :WIDTH 0 :HEIGHT 0 :MIPMAPS 0 :DATA-FORMAT 0 #<RL-TEXTURE {1009735893}>> :RECS #<unbound-slot> :GLYPHS #<unbound-slot> :PATH ~/.guix-profile/share/fonts/truetype/Cormorant-Bold.ttf :ASSET #<RL-FONT :TEXTURE #<RL-TEXTURE :ID 0 :WIDTH 0 :HEIGHT 0 :MIPMAPS 0 :DATA-FORMAT 0 #<RL-TEXTURE {1009735893}>> :RECS #<unbound-slot> :GLYPHS #<unbound-slot> :SIZE 0 :GLYPH-COUNT 0 :GLYPH-PADDING 0 #<RL-FONT {10096E0BC3}>> >`